pkg/controller/container-runtime-config/helpers.go

```diff
@@ -495,29 +496,64 @@ func validateUserContainerRuntimeConfig(cfg *mcfgv1.ContainerRuntimeConfig) erro
+func getValidBlockedRegistries(releaseImage string, imgSpec *apicfgv1.ImageSpec, icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) ([]string, error) {
 	if imgSpec == nil {
 		return nil, nil
 	}
 
-	var blockedRegs []string
+	var (
+		blockedRegs []string
+		blockErr    []string
+	)
 
 	// Get the registry being used by the payload from the releaseImage
 	ref, err := reference.ParseNamed(releaseImage)
 	if err != nil {
 		return nil, errParsingReference
 	}
-	payloadReg := reference.Domain(ref)
-	for i, reg := range imgSpec.RegistrySources.BlockedRegistries {
+	payloadRepo := reference.TrimNamed(ref).String()
+	for _, reg := range imgSpec.RegistrySources.BlockedRegistries {
 		// if there is a match, return all the blocked registries except the one that matched and return an error as well
-		if reg == payloadReg {
-			blockedRegs = append(blockedRegs, imgSpec.RegistrySources.BlockedRegistries[i+1:]...)
-			return blockedRegs, fmt.Errorf("error adding %q to blocked registries, cannot block the registry being used by the payload", payloadReg)
+		if runtimeutils.ScopeIsNestedInsideScope(payloadRepo, reg) {
+			// If the payload registry doesn't have mirror rules configured for it, then don't add it to the blocked registries list
+			if !payloadRegHasMirrorConfig(payloadRepo, icspRules, imgSpec.RegistrySources.BlockedRegistries) {
+				blockErr = append(blockErr, reg)
+				continue
+			}
+			// Log a warning that we are adding the payload registries to the blocked registries list as there are mirror rules for it
+			glog.Warningf("%q is the payload registry, but will add it to the list of blocked registries as there are mirror rules configured for it", reg)
+
 		}
 		// Was not a match to the registry being used by the payload, so add to valid blocked registries
 		blockedRegs = append(blockedRegs, reg)
 	}
-	return blockedRegs, nil
+	var retErr error
+	if len(blockErr) > 0 {
+		retErr = fmt.Errorf("error adding %q to blocked registries, cannot block the registry being used by the payload", blockErr)
+	}
+	return blockedRegs, retErr
+}
+
+// payloadRegHasMirrorConfig returns true if the payload registry has mirror rules configured for it
+func payloadRegHasMirrorConfig(payloadRepo string, icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy, blockedRegs []string) bool {
+	var payloadMirrors []string
+	// Go through the mirror rules configured and check if a source matches the payload registry
+	for _, rule := range icspRules {
+		for _, mirror := range rule.Spec.RepositoryDigestMirrors {
+			if runtimeutils.ScopeIsNestedInsideScope(payloadRepo, mirror.Source) {
+				payloadMirrors = append(payloadMirrors, mirror.Mirrors...)
+			}
+		}
+	}
+	// Go through the list of mirrors and check that at least one of the payload mirrors is not blocked
+	for _, mirror := range payloadMirrors {
+		for _, blockReg := range blockedRegs {
+			if !runtimeutils.ScopeIsNestedInsideScope(mirror, blockReg) {
```

Comment:
This terminates on the first blocking entry that doesn’t match a mirror — but there might be another blocking entry that matches that mirror. Shouldn’t this succeed only if we find a mirror that his blocked by none of `blockedRegs`?

Reply:
fixed